theseus/core/cost_function.py

```diff
@@ -263,11 +272,20 @@ def _expand_all(
     def jacobians(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
         err, optim_vars, aux_vars = self._compute_error()
         if self._autograd_mode == AutogradMode.VMAP:
-            jacobians_full = self._compute_autograd_jacobian_vmap(
-                tuple(v.tensor for v in optim_vars),
-                tuple(v.tensor for v in aux_vars),
-                self._make_jac_fn_vmap(self._tmp_optim_vars, self._tmp_aux_vars),
-            )
+            # functorch doesn't allow BatchedTensors created inside their transforms
+            # to be accessed outside that context. Since our tmp containers get
+            # populated inside the vmap(jacrev) calculation, this would result in
+            # errors when calling functions like copy() or to() after
+            # vmap has been run at least once.
+            # The _tmp_tensors context managers stores the (regular) tensors the vars
+            # had before entering and restores them on exit, thus dereferencing
+            # the temporary BatchedTensors.
+            with _tmp_tensors(self._tmp_optim_vars), _tmp_tensors(self._tmp_aux_vars):
```

Comment:
This is a nice fix! There is likely very little overhead I am assuming?

Reply:
So functorch essentially changes the types of `self._tmp_optim_vars` and `self._tmp_aux_vars`?

Reply:
Looks like functorch's vmap replaces all tensor inputs with copies of a class called `BatchedTensor` and returns tensors of this type that get converted to regular tensors when the function returns. If you have some global reference to a `BatchedTensor` created inside but never returned (e.g., by storing in `self._tmp_optim_vars`), then it never gets converted to a regular tensor and we get an error if we try to access it outside.

Reply:
@mhmukadam I would assume that the overhead of this is not too large compared to the costly operation will happen inside, but I can run some checks just in case.

Reply:
> Looks like functorch's vmap replaces all tensor inputs with copies of a class called `BatchedTensor` and returns tensors of this type that get converted to regular tensors when the function returns. If you have some global reference to a `BatchedTensor` created inside but never returned (e.g., by storing in `self._tmp_optim_vars`), then it never gets converted to a regular tensor and we get an error if we try to access it outside.

Great catch! Not sure if this will raise other bugs since the reference types are essentially changed. BTW, what is the type of `jacobian`? Is it a regular `Tensor`?

Reply:
Yeah, `jacobian` should be a regular tensor, since we get it as the return value from `vmap`. 

Looking at the code below, the only potential source of errors I can see is if `_err_fn` stores internal data during the error computation. We should maybe document that `err_fn` should be a [pure function](https://en.wikipedia.org/wiki/Pure_function). 

<img width="756" alt="image" src="https://user-images.githubusercontent.com/4759586/197634371-1c95d8a7-4564-4673-b633-871548286ef3.png">

Reply:
I'll add a comment at the top of `AutoDiffCostFunction`. Also, my last statement wasn't correct. I think it's fine if the function uses some outside information to compute error (e.g., some model that gets updated periodically), but what's not allowed is that the function stores the tensors used as input to be reused elsewhere.

Reply:
Added comment. Should I merge? I ran some small tests locally and I didn't notice any overhead.